src/datachain/lib/dc.py

```diff
@@ -331,22 +346,80 @@ def from_storage(
             chain = DataChain.from_storage("s3://my-bucket/my-dir")
             ```
         """
-        func = get_file(type)
-        return (
-            cls(
-                path,
+        file_type = get_file_type(type)
+
+        if client_config is None:
+            client_config = {}
+
+        if anon:
+            client_config["anon"] = True
+
+        session = Session.get(
+            session,
+            catalog=kwargs.get("catalog"),
+            client_config=client_config,
+            in_memory=in_memory,
+        )
+
+        client, path = Client.parse_url(
+            uri, session.catalog.cache, **session.catalog.client_config
+        )
+
+        # clean path without globs
+        lst_path = (
+            posixpath.dirname(path)
+            if glob.has_magic(os.path.basename(os.path.normpath(path)))
+            or client.fs.isfile(uri)
+            else path
+        )
+        lst_uri = f"{client.uri}/{lst_path.lstrip('/')}"
+
+        ds_name = listing_dataset_name(client.uri, posixpath.join(lst_path, ""))
+
+        for ds in cls.datasets(
+            session=session, in_memory=in_memory, include_listing=True
+        ).collect("dataset"):
+            if listing_expired(ds.created_at):  # type: ignore[union-attr]
+                continue
+            if (
+                listing_subset(ds.name, ds_name)  # type: ignore[union-attr]
+                and not update
+            ):
+                # we can use found listing as it contains the one from input
+                dc = cls.from_dataset(ds.name, **kwargs)  # type: ignore[union-attr]
+                dc.signals_schema = dc.signals_schema.mutate(
```

Comment:
Q: why is this needed?

Reply:
When we list bucket, listing dataset that holds data will always have generic `File` in schema, e.g `{"file": "File@v1"}`, but we need to be able to return chain from `.from_storage()` that is using different type, e.g `TextFile` which is done with `type` argument.
Example:

```python
txt_dc = DataChain.from_storage("gs://datachain-demo", type="text").filter(C("file.path").glob("*.txt")) 
```


So that line of code just changes type of object holding listing, e.g `File` -> `TextFile` in the schema.